backend/substrapp/tasks/k8s_backend.py

```diff
@@ -446,6 +448,102 @@ def k8s_remove_image(image_name):
         logger.exception(e)
 
 
+def k8s_get_image_list():
+    response = requests.get(
+        f'{REGISTRY_SCHEME}://{REGISTRY}/v2/_catalog',
+        headers={'Accept': 'application/vnd.docker.distribution.manifest.v2+json'},
+        timeout=HTTP_CLIENT_TIMEOUT_SECONDS
+    )
+
+    response.raise_for_status()
+
+    images = []
+    res = response.json()
+    for repository in res['repositories']:
+        response = requests.get(
+            f'{REGISTRY_SCHEME}://{REGISTRY}/v2/{repository}/tags/list',
+            headers={'Accept': 'application/vnd.docker.distribution.manifest.v2+json'},
+            timeout=HTTP_CLIENT_TIMEOUT_SECONDS
+        )
+
+        response.raise_for_status()
+        images.append(response.json())
+
+    return images
+
+
+def k8s_fetch_old_algo_image_names(max_duration):
+
+    images = [image for image in k8s_get_image_list() if image['tags'] and 'substra' in image['tags']]
+
+    old_images = []
+
+    for image in images:
+        response = requests.get(
+            f'{REGISTRY_SCHEME}://{REGISTRY}/v2/{image["name"]}/manifests/substra',
+            timeout=HTTP_CLIENT_TIMEOUT_SECONDS
+        )
+
+        response.raise_for_status()
+
+        # take the most recent date as creation date
+        created_date = max([
+            datetime.datetime.strptime(json.loads(e['v1Compatibility'])['created'].split('.')[0],
+                                       '%Y-%m-%dT%H:%M:%S')
+            for e in response.json()['history']
+        ])
+
+        if (datetime.datetime.now() - created_date).total_seconds() >= max_duration:
+            old_images.append(image["name"])
+
+    return old_images
+
+
+def k8s_docker_registry_garbage_collector():
+    kubernetes.config.load_incluster_config()
+    k8s_client = kubernetes.client.CoreV1Api()
+    pod_name = get_pod_name('docker-registry')
+    exec_command = [
+        '/bin/sh',
+        '-c',
+        '/bin/registry garbage-collect /etc/docker/registry/config.yml'
+    ]
+
+    resp = kubernetes.stream.stream(
+        k8s_client.connect_get_namespaced_pod_exec,
+        pod_name,
+        NAMESPACE,
+        command=exec_command,
+        stderr=True,
+        stdin=True,
+        stdout=True,
+        tty=False,
+        _preload_content=False,
+    )
+
+    logs = []
+
+    while resp.is_open():
+        lines = ""
+        resp.update(timeout=1)
+        if resp.peek_stdout():
+            lines = resp.read_stdout()
+        if resp.peek_stderr():
+            lines = resp.read_stderr()
+        for line in filter(None, lines.split("\n")):
+            logs.append(line)
+    else:
+        logger.info(logs[-1])
```

Comment:
Note that if stdout and stderr are written to within the same second (and in this case only), the corresponding lines might not be in the right order (stdout might be before stderr even though stderr was printed before).

The way to fix this is to use tty=True instead, and to redirect stderr to stdout `2>&1`